engine/src/witness/btc/smart_contract.rs

```diff
@@ -0,0 +1,295 @@
+use cf_chains::{
+	btc::{smart_contract_encoding::UtxoEncodedData, ScriptPubkey},
+	ForeignChainAddress,
+};
+use cf_primitives::{Asset, AssetAmount, Price};
+use codec::Decode;
+use itertools::Itertools;
+use utilities::SliceToArray;
+
+use crate::btc::rpc::VerboseTransaction;
+
+const OP_PUSHBYTES_75: u8 = 0x4b;
+const OP_PUSHDATA1: u8 = 0x4c;
+
+#[derive(PartialEq, Debug)]
+pub struct BtcContractCall {
+	output_asset: Asset,
+	deposit_amount: AssetAmount,
+	output_address: ForeignChainAddress,
+	// --- FoK ---
+	retry_duration: u16,
+	refund_address: ScriptPubkey,
+	min_price: Price,
+	// --- DCA ---
+	number_of_chunks: u16,
+	chunk_interval: u16,
+	// --- Boost ---
+	boost_fee: u8,
+}
+
+fn try_extract_utxo_encoded_data(script: &bitcoin::ScriptBuf) -> Option<&[u8]> {
+	let bytes = script.as_script().as_bytes();
+
+	// First opcode must be OP_RETURN
+	if bytes[0] != OP_RETURN.to_u8() {
+		return None;
+	}
+
+	// Second opcode must be either OP_PUSHBYTES_X (1..=75) or OP_PUSHDATA1 (76):
+	let (data_len, data_bytes) = match bytes[1] {
+		// Opcode encodes the length directly
+		data_len @ 1..=OP_PUSHBYTES_75 => (data_len, &bytes[2..]),
+		// The length is encoded in the following byte:
+		OP_PUSHDATA1 => (bytes[2], &bytes[3..]),
+		_ => {
+			return None;
+		},
+	};
+
+	// Sanity check:
+	if data_bytes.len() != data_len as usize {
+		return None;
+	}
+
+	Some(data_bytes)
+}
+
+fn script_buf_to_script_pubkey(script: &ScriptBuf) -> Option<ScriptPubkey> {
+	fn data_from_script<const LEN: usize>(script: &ScriptBuf, bytes_to_skip: usize) -> [u8; LEN] {
+		script.bytes().skip(bytes_to_skip).take(LEN).collect_vec().as_array()
+	}
+
+	let pubkey = if script.is_p2pkh() {
+		ScriptPubkey::P2PKH(data_from_script(script, 3))
+	} else if script.is_p2sh() {
+		ScriptPubkey::P2SH(data_from_script(script, 2))
+	} else if script.is_v1_p2tr() {
+		ScriptPubkey::Taproot(data_from_script(script, 2))
+	} else if script.is_v0_p2wsh() {
+		ScriptPubkey::P2WSH(data_from_script(script, 2))
+	} else if script.is_v0_p2wpkh() {
+		ScriptPubkey::P2WPKH(data_from_script(script, 2))
+	} else {
+		ScriptPubkey::OtherSegwit {
+			version: script.witness_version()?.to_num(),
+			program: script.bytes().skip(2).collect_vec().try_into().ok()?,
+		}
+	};
+
+	Some(pubkey)
+}
+
+// Currently unused, but will be used by the deposit wintesser:
+#[allow(dead_code)]
+pub fn try_extract_contract_call(
+	tx: &VerboseTransaction,
+	vault_address: ScriptPubkey,
+) -> Option<BtcContractCall> {
+	// A correctly constructed transaction carrying CF swap parameters must have at least 3 outputs:
+	let [utxo_to_vault, nulldata_utxo, change_utxo, ..] = &tx.vout[..] else {
+		return None;
+	};
+
+	// First output must be a deposit into our vault:
+	if utxo_to_vault.script_pubkey.as_bytes() != vault_address.bytes() {
+		return None;
+	}
+
+	// Second output must be a nulldata UTXO (with 0 amount):
+	if nulldata_utxo.value.to_sat() != 0 {
+		return None;
```

Comment:
A warning or debug statement might be appropriate here too.

Reply:
I'm adding a warning, but are we not expecting other transactions into our vault's change address? Though I guess if we are sending change to ourselves it will never be the first UTXO.

Reply:
Yeah I was just thinking in terms of debugging that this it might help. No big deal though.